### DISCUSSION TITLE
Automatic DOM diffing

### DISCUSSION THREAD

**Discussion by kpietraszko** [unpoly/unpoly#417] [Ideas]
Hi. Sorry if this was discussed or implemented already, couldn't find it.
I'd love to see an automatic DOM diffing mechanism when targeting `up-main`. I mean something along the lines of what React does with its virtual DOM; or another solution like [morphdom](https://github.com/patrick-steele-idem/morphdom) , which is used by Pheonix LiveWire and Django Sockpuppet.

I think it would be extremely convenient, without worrying about placing your fragments, writing selectors etc. What are your thoughts?

**Comment by kpietraszko** [comment]
Seems like _htmx_ has that feature as an extension: https://htmx.org/extensions/morphdom-swap/ or https://htmx.org/extensions/alpine-morph/

**Reply by kpietraszko** [reply]
There's also Google's [incremental-dom](https://github.com/google/incremental-dom)

**Comment by triskweline** [comment]
I find this very interesting, but needs some deep thinking what the impact would be.

I know that the authors of htmx and Alpine.js/LiveWire are both working on new DOM morphing libraries because they ran into limitations with morphdom:

https://github.com/bigskysoftware/idiomorph
https://github.com/alpinejs/alpine/blob/main/packages/morph/src/morph.js

**Reply by kpietraszko** [reply]
Yeah Alpine has been using their own solution for some time now, morphdom doesn't even seem to be updated anymore. But even their solution is far from perfect and not really intended for full-page morphing. There's a PR fixing that, otherwise things like `<head>` get wonky.
In general I find the html-over-the-write approach (LiveWire, RoR's HotWire, even ASP.NET Blazor Server) fascinating. I'd love a way to use any backend technology and perfectly morph into the returned HTML, totally abstracted away.

**Reply by kpietraszko** [reply]
Speaking of the devil, this happened yesterday: https://twitter.com/htmx_org/status/1564767845640183809

**Reply by davisums** [reply]
Looks like Turbo 8 also went with idiomorph because of issues with morphdom and they seem happy with it (mentioned on the idiomorph homepage).

**Comment by thetutlage** [comment]
Coming here after Turbo added morphing recently. One place, where I really like morphing is with form submissions, as the user submitting the form does not loose the input values and more importantly the focus from the input field.

Would love to see something similar coming to Unpoly :)

**Comment by hpvd** [comment]
just to add other POVs/ideas (without rating):

today there are also liveview/livewire projects have already switched away from dom diff for some time.

This is a small, but really advanced project with pretty interesting features and some really deep thoughts and considerations inside (also in other fields beside dom diff):

**ts-liveview**
https://github.com/beenotung/ts-liveview/?tab=readme-ov-file#ts-liveview
DEMO: https://liveviews.cc/

of course, it's **not directly comparable** from the taken approach/goal side and not as battery included as unpoly, but the result is really stunning small and fast:
https://github.com/beenotung/ts-liveview/tree/master?tab=readme-ov-file#size-comparison-with-other-tools
![2024-02-12_12h47_15](https://github.com/unpoly/unpoly/assets/5681880/7592cec5-1bab-4eb7-904c-f44e73449ab7)


It started with morphdom and moves on 2 years ago, see changelog of v2:
https://github.com/beenotung/ts-liveview/blob/master/CHANGELOG.md
